File: scripts/remove_SNPs_within_SVs.py
#!/usr/bin/python3

'''
version 1.0

created by Rafael Della Coletta
2019-09-26
'''


import argparse as ap


# initialize argument parser (pass user input from command line to script)
parser = ap.ArgumentParser(formatter_class=ap.RawDescriptionHelpFormatter,
                           description='''
description: this script removes SNPs from a VCF file that fall within the
             boundaries of a structural variant in another VCF file.

note: VCF header from original and filtered files are the same.''')
# add positional arguments
parser.add_argument("vcf_SNPs", type=str,
                    help="VCF file with SNPs calls")
parser.add_argument("vcf_SVs", type=str,
                    help="VCF file with structural variant calls")
parser.add_argument("output_name", type=str,
                    help="name of output file")
# optional argument for which type of variants to exclude?

# pass arguments into variables
args = parser.parse_args()
vcf_SNPs = args.vcf_SNPs
vcf_SVs = args.vcf_SVs
output_name = args.output_name


# funtion to extract positions of SNPs or SVs
def get_variant_info(vcf_file, is_SV=False):

    # define which information to extract from VCF header
    info_list = ["CHROM", "POS", "INFO"]
    # create list to store indices of each info above
    header_idx = []
    # create dictionary to store information extracted
    variant_info = {}

    # read vcf line by line
    for line in vcf_file:
        line = line.strip()
        # if line starts with ## -- skip
        if line[0:2] == "##":
            continue
        # if line starts with # -- save as header
        elif line[0:1] == "#":
            # save header
            header = line.split("\t")
            # remove # from #CHROM
            header[0] = header[0][1:]
            # get indices of columns with important information from header
            for info in info_list:
                header_idx.append(header.index(info))
            # print(header_idx)
        # otherwise, extract info from line
        else:
            line = line.split("\t")
            # get chrom number
            chr = line[header_idx[0]]
            # get snp position
            pos = line[header_idx[1]]

            # if vcf file is from SNPs...
            if is_SV is False:
                # add coordinates to dictionary
                if chr not in variant_info:
                    variant_info[chr] = [pos]
                else:
                    variant_info[chr].append(pos)

            # if vcf file is from SVs...
            if is_SV is True:
                # get sv type
                sv_type = line[header_idx[2]].split(";SVTYPE=")
                sv_type = sv_type[1].split(";")[0]
                # exclude translocations
                if sv_type != "TRA":
                    # get end position of sv
                    sv_end = line[header_idx[2]].split(";END=")
                    sv_end = sv_end[1].split(";")[0]
                    # add sv info to dictionary
                    sv_info = pos + "," + sv_end + "," + sv_type
                    if chr not in variant_info:
                        variant_info[chr] = [sv_info]
                    else:
                        variant_info[chr].append(sv_info)

    return variant_info


# open input files
infile_SNPs = open(vcf_SNPs, "r")
infile_SVs = open(vcf_SVs, "r")

# parse vcf file to get coordinates of variants
print("Extracting SNPs and SVs coordinates...")

SNPs_info = get_variant_info(infile_SNPs, is_SV=False)
SVs_info = get_variant_info(infile_SVs, is_SV=True)
print("Done!")

# close files
infile_SNPs.close()
infile_SVs.close()


# the following code is a bit cumbersome, but it's a quicker way to find which
# SNPs are in the range of a SV than using a bunch of loops inside loops


# create a dictionary to store all bases that are in a SV (e.g., if SV ranges
# from position 100 to 105, I will store the bases 100,101,102,103,104,105)
SVs_range = {}
# create list to store positions of SNPs to be excluded from analysis
SNPs_within_SVs = {}

print("\nIdentifying SNPs within a SV...")

# for each SV in each chromosome
for chr in SVs_info.keys():
    for SV in SVs_info[chr]:
        # get SV range
        sv_start = int(SV.split(",")[0])
        sv_end = int(SV.split(",")[1])
        sv_range = list(range(sv_start, sv_end + 1, 1))
        # add all bases of the SV range into the dictionary
        if chr not in SVs_range:
            SVs_range[chr] = [sv_range]
        else:
            # SVs_range[chr] will be a list with many sublists inside
            SVs_range[chr].append(sv_range)
    # transform all sublists into one flat list
    SVs_range[chr] = [base for range in SVs_range[chr] for base in range]
    # remove redundancy
    SVs_range[chr] = set(SVs_range[chr])
    # keep only SNPs that are within the range
    if chr not in SNPs_within_SVs:
        SNPs_within_SVs[chr] = [SNP for SNP in SNPs_info[chr] if int(SNP) in SVs_range[chr]]
    else:
        [SNPs_within_SVs[chr].append(SNP) for SNP in SNPs_info[chr] if int(SNP) in SVs_range[chr]]
    # erase all values for that dictionary key to save memory
    SVs_range[chr] = 0
    print("  ", len(SNPs_within_SVs[chr]), "SNPs within a SV in", chr + "!")


# open SNP vcf file again
infile_SNPs = open(vcf_SNPs, "r")

# also open output file to write the filtered vcf
outfile = open(output_name, "w")

print("\nGenerating new VCF file only with SNPs not within a SV range...")

# read vcf line by line
for line in infile_SNPs:
    line = line.strip()
    # if line starts with ## -- print line to new vcf
    if line[0:1] == "#":
        print(line, file=outfile)
    # otherwise, extract info from line
    else:
        # get chrom number and snp position
        chr = line.split("\t")[0]
        pos = line.split("\t")[1]
        # make sure the chromosome being parsed has SNPs within a SV
        if chr in SNPs_within_SVs.keys():
            # only print lines to new vcf if positions are not within a SV
            if pos not in SNPs_within_SVs[chr]:
                print(line, file=outfile)

print("Done!")

# close files
infile_SNPs.close()
outfile.close()

File: scripts/remove_SNPs_within_SVs.R
#### arguments for command line ----

args <- commandArgs(trailingOnly = TRUE)

# help
if (all(length(args) == 1 & args == "-h" | args == "--help")) {
  cat("
Description: this script filter SNPs out of a hapmap file based on coordinates of SNPs
             (chromosome and position) in another tab-delimited file.

Usage: Rscript remove_SNPs_within_SVs [hmp_file] [file_with_SNPs_coord] [output_name]\n\n")
  quit()
}

# make sure the correct number of arguments are used
# you should provide 3 arguments
if (length(args) != 3) {
  stop("incorrect number of arguments provided.
  
Usage: Rscript remove_SNPs_within_SVs [hmp_file] [file_with_SNPs_coord] [output_name]
       ")
}

# assign arguments to variables
hmp.file <- args[1]
snps.file <- args[2]
output.name <- args[3]



#### libraries ----

if(!require("data.table")) install.packages("data.table")



#### filtering ----

# load data
cat("Loading data...\n")
hmp.file <- fread(hmp.file, header = TRUE, data.table = FALSE)
snps.file <- fread(snps.file, header = FALSE, data.table = FALSE)
cat("Done!\n\n")

# get chromosomes in the hmp file
chrs.to.parse <- unique(hmp.file[, "chrom"])

# create an empty data frame to store output
output.hmp <- data.frame(matrix(nrow = 0, ncol = NCOL(hmp.file)))
colnames(output.hmp) <- colnames(hmp.file)


for (chr in chrs.to.parse) {
  
  cat("Removing SNPs from chr ", chr, "...\n", sep = "")
  
  # make sure snps to remove are in the same chr as the ones in the hapmap
  snps.to.remove <- snps.file[which(snps.file[, 1] == chr), 2]
  hmp.file.filtered <- hmp.file[which(hmp.file[, "chrom"] == chr), ]
  
  # remove SNPs
  hmp.file.filtered <- hmp.file.filtered[which(!hmp.file.filtered[, "pos"] %in% snps.to.remove), ]
  
  # append to final hapmap
  output.hmp <- rbind(output.hmp, hmp.file.filtered)
  
  cat("Done!\n\n")
}


# count how many SNPs were removed
for (chr in chrs.to.parse) {
  
  if (chr %in% unique(snps.file[, 1])) {
    n.snps.before <- NROW(hmp.file[which(hmp.file[, "chrom"] == chr), ])
    n.snps.after <- NROW(output.hmp[which(output.hmp[, "chrom"] == chr), ])
    n.snps.removed <- n.snps.before - n.snps.after
    
    cat(n.snps.removed, "SNPs removed in chromosome", chr, "\n")
  }
}

# write output
fwrite(output.hmp, file = output.name, sep = "\t", quote = FALSE, na = "NA")
